pkg/storage/internalstorage/storage.go

```diff
@@ -11,19 +13,71 @@ import (
 	"github.com/clusterpedia-io/clusterpedia/pkg/storage"
 )
 
+var mutex sync.Mutex
+
 type StorageFactory struct {
-	db *gorm.DB
+	db             *gorm.DB
+	AutoMigration  *bool
+	DivisionPolicy DivisionPolicy
+	Mapper         []ResourceMapper
+}
+
+func (s *StorageFactory) AutoMigrate() error {
+	return nil
 }
 
 func (s *StorageFactory) GetSupportedRequestVerbs() []string {
 	return []string{"get", "list"}
 }
 
 func (s *StorageFactory) NewResourceStorage(config *storage.ResourceStorageConfig) (storage.ResourceStorage, error) {
-	return &ResourceStorage{
-		db:    s.db,
-		codec: config.Codec,
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	gvr := schema.GroupVersionResource{
+		Group:    config.StorageGroupResource.Group,
+		Version:  config.StorageVersion.Version,
+		Resource: config.StorageGroupResource.Resource,
+	}
+	table := s.tableName(gvr)
+
+	var model interface{}
+	if s.AutoMigration != nil && *s.AutoMigration {
```

Comment:
Perhaps it would be clearer to categorize by `s.DivisionPolicy` first, eg.
```go
        model := &Resource{}
        switch s.DivisionPolicy {
        case DivisionPolicyGroupResource:
                model = &GroupVersionResource{}

                if !s.SkipAutoMigration {
                        if exist := s.db.Migrator().HasTable(table); !exist {
                                if err := s.db.AutoMigrate(model); err != nil {
                                        return nil, err
                                }

                                if err := s.db.Migrator().RenameTable("groupversionresources", table); err != nil {
                                        return nil, err
                                }
                        }
                }
        }
```